src/document.rs

```diff
@@ -170,7 +168,7 @@ fn extract_value_single_or_list_for_type(
                 .unwrap_or(false)
         {
             return Ok(vec![extract_value_for_type(
-                values,
+                values.into_any(),
```

Comment:
I am not yet sure what these calls like `into_any` do. I added this because it seemed correct and made the compiler happy, but not sure if there's a better way to resolve these situations.

Reply:
`into_any` just erases the type information, i.e. it is basically an upcast and the converse to `downcast`. It is necessary because Rust does not have a notion of how `Bound<'_, T>` and `Bound<'_, S>` are related based on how `T` and `S`. In our case, `T` are usually subtypes of `PyAny` and hence the `into_any` upcast is available.

In the past, this was "automatic" because `&'py PyDict` allow getting a short-lived `&'a PyAny` via the `Deref` trait which the compiler calls automatically. We still have [that `Deref`](https://docs.rs/pyo3/latest/pyo3/struct.Bound.html#impl-Deref-for-Bound%3C'py,+T%3E) but it only goes from `Bound<'py, T>` to `&'a Bound<'py, PyAny>` which is insufficient here.

I would say, this is another example where the old PyO3 API conflated the short-lived borrow `'a` and the GIL lifetime `'py` and it somehow worked out via covariance. This was nice, but again the price for this convenience was mandatory global state tracking making everyone's Python-to-Rust calls slower.